types/wlr_virtual_keyboard.c

```diff
@@ -0,0 +1,208 @@
+		uint32_t mods_latched, uint32_t mods_locked, uint32_t group) {
+	struct wlr_virtual_keyboard *keyboard =
+		virtual_keyboard_from_resource(resource);
+	wlr_keyboard_notify_modifiers(keyboard->input_device.keyboard,
+		mods_depressed, mods_latched, mods_locked, group);
+}
+
+static void virtual_keyboard_destroy_resource(struct wl_resource *resource) {
+	struct wlr_virtual_keyboard *keyboard =
+		virtual_keyboard_from_resource(resource);
+	wlr_signal_emit_safe(&keyboard->events.destroy, keyboard);
+	wlr_keyboard_destroy(keyboard->input_device.keyboard);
+	free(keyboard);
+}
+
+static void virtual_keyboard_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	virtual_keyboard_destroy_resource(resource);
+}
+
+static const struct zwp_virtual_keyboard_v1_interface virtual_keyboard_impl = {
+	.keymap = virtual_keyboard_keymap,
+	.key = virtual_keyboard_key,
+	.modifiers = virtual_keyboard_modifiers,
+	.destroy = virtual_keyboard_destroy,
+};
+
+static const struct zwp_virtual_keyboard_manager_v1_interface manager_impl;
+
+static struct wlr_virtual_keyboard_manager *manager_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_virtual_keyboard_manager_v1_interface, &manager_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_keyboard_manager_create_virtual_keyboard(
+		struct wl_client *client, struct wl_resource *resource,
+		struct wl_resource *seat,
+		uint32_t id) {
+	struct wlr_virtual_keyboard_manager *manager =
+		manager_from_resource(resource);
+
+	struct wlr_virtual_keyboard *virtual_keyboard = calloc(1,
+		sizeof(struct wlr_virtual_keyboard));
+	if (!virtual_keyboard) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_keyboard* keyboard = calloc(1, sizeof(struct wlr_keyboard));
+	if (!keyboard) {
+		wlr_log(L_ERROR, "Cannot allocate wlr_keyboard");
+		free(virtual_keyboard);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wlr_keyboard_init(keyboard, &keyboard_impl);
+
+	struct wl_resource *keyboard_resource = wl_resource_create(client,
+		&zwp_virtual_keyboard_v1_interface, wl_resource_get_version(resource),
+		id);
+	if (!keyboard_resource) {
+		free(keyboard);
+		free(virtual_keyboard);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(keyboard_resource, &virtual_keyboard_impl,
+		virtual_keyboard, virtual_keyboard_destroy_resource);
+
+	wlr_input_device_init(&virtual_keyboard->input_device,
+		WLR_INPUT_DEVICE_KEYBOARD, &input_device_impl, "virtual keyboard",
+		0xabcd, 0x1234);
+
+	struct wlr_seat_client *seat_client = wlr_seat_client_from_resource(seat);
+
+	virtual_keyboard->input_device.keyboard = keyboard;
+	virtual_keyboard->resource = keyboard_resource;
+	virtual_keyboard->seat = seat_client->seat;
+	wl_signal_init(&virtual_keyboard->events.destroy);
+	wlr_signal_emit_safe(&manager->events.new_virtual_keyboard, virtual_keyboard);
```

Comment:
We should have a list of all virtual keyboards in the manager.

Reply:
What for?

Reply:
Let the users iterate on the list and properly cleanup when destroying the manager.

Reply:
What's the advantage of letting manager do the cleanup over letting keyboards clean up themselves?
As I understand, there's not much utility in destroying the manager while objects are bound. In addition to that, the manager is acting as a factory rather than a collection, so there's no harm in leaving keyboards alive after destroying it.
In addition to that, having a list of active keyboards in the manager will need introducing another object to remember which keyboards are held by which user, making the whole thing a doubly-nested structure:

- manager global
- manager binds
- keyboards

The way it is now seems just as functional and can stay simple.

Reply:
Giving it a bit more thought, a binds list is not implied by a keyboard list, but I stand by my comment that it's not necessary.

Reply:
No need for complicated list. The global, a list of resources and a list of keyboards in the manager is enough.

The list is useful to let compositors inspect it without having to maintain their own list.